2-onchain-interactions.md

```diff
@@ -0,0 +1,284 @@
+};
+
+// TODO: why not make these the same as the contract event names where
+// applicable?
+type ChainEventName =
+  | "fundsDeposited"
+  | "fundsWithdrawn"
+  | "transactionMined"
+  | "forceMoveDetected"
+  | "checkpointDetected"
+  | "concludeDetected"
+  | "respondDetected";
+
+type ChainEvent = {
+  channelId: Bytes32;
+  event: {
+    type: ChainEventName;
+    // Example data type
+    // https://github.com/statechannels/statechannels/blob/master/packages/nitro-protocol/src/contract/asset-holder.ts#L5
+    data: ContractEvent;
+  };
+};
+
+// TODO: finalize this! how aware of the wallet data structure is the onchain service expected to be?
+// instead, should this directly mirror the onchain state and be responsible for constructing the state representation
+// defined by the asset holder contracts instead?
+type ChainState = any;
+
+export interface IOnchainService {
+  constructor(config: OnchainServiceEnvironment);
+
+  /**
+   * @description Static method to create and begin syncing an onchain service
+   * @param config The environment for the service
+   * @returns A newly started and syncing instance of the onchain service
+   *
+   * @notice Not implemented until v2, may not be implemented at all (mostly included as discussion)
+   */
+  public static start(
+    config: OnchainServiceEnvironment
+  ): Promise<IOnchainService>;
+
+  /**
+   * @description Stops the onchain service from monitoring the chain
+   * @returns A boolean indicating success
+   *
+   * @notice Not implemented until v2, depends on how historical syncing is done
+   */
+  // TODO: see below for note about booleans
+  public stop(): Promise<boolean>;
+
+  /**
+   * @description Connects the wallet to the given provider
+   * @param provider The provider to connect to
+   * @returns The updated service
+   */
+  public connect(
+    provider: string | providers.JsonRpcProvider
+  ): Promise<IOnchainService>;
+
+  /**
+   * @description Registers a channel for the onchain service to watch for.
+   * @param channelId Unique channel identifier
+   * @param assetHolders Asset holder contract addresses
+   * @returns boolean value indicating if registration was successful
+   */
+  // TODO: booleans for success are rarely very informative, since you get little
+  // insight into why the failure happened. I would suggest returning a string value
+  // instead
+  public registerChannel(
+    channelId: Bytes32,
+    assetHolders: AssetHolderAddress[]
+  ): Promise<boolean>;
+
+  /**
+   * @description Submits a transaction to chain
+   * @param tx Minimum transaction to send
+   * @param completionEvent An event you expect to be emitted when the transaction can be
+   * considered "completed" from the POV of the chain wallet (i.e. all data stored, see TODO
+   * above if this is even necessary)
+   * @returns A transaction result
+   */
+  public submitTransaction(
+    tx: MinimumTransaction,
+    completionEvent: ChainEventName
+  ): Promise<TransactionResult>;
```

Comment:
Is this transaction signed beforehand or not? If not, where does it get signed? What does nonce management?